nni/algorithms/nas/pytorch/fbnet/trainer.py

```diff
@@ -0,0 +1,417 @@
+                labels = labels.to(self.device, non_blocking=True)
+
+                if self.arch_search:
+                    perf_cost = Variable(torch.zeros(self.dev_num, 1)).to(
+                        self.device, non_blocking=True
+                    )
+                    output, _ = self.model(images, self.temp, perf_cost)
+
+                else:
+                    output = self.model(images)
+
+                loss = self.criterion(output, labels)
+                acc1, acc5 = accuracy(output, labels, topk=(1, 5))
+                losses.update(loss, images.size(0))
+                top1.update(acc1[0], images.size(0))
+                top5.update(acc5[0], images.size(0))
+                # measure elapsed time
+                batch_time.update(time.time() - end)
+                end = time.time()
+
+                if i % 10 == 0 or i + 1 == len(self.valid_loader):
+                    test_log = (
+                        "Valid" + ": [{0}/{1}]\t"
+                        "Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t"
+                        "Loss {loss.val:.4f} ({loss.avg:.4f})\t"
+                        "Top-1 acc {top1.val:.3f} ({top1.avg:.3f})\t"
+                        "Top-5 acc {top5.val:.3f} ({top5.avg:.3f})".format(
+                            i,
+                            len(self.valid_loader) - 1,
+                            batch_time=batch_time,
+                            loss=losses,
+                            top1=top1,
+                            top5=top5,
+                        )
+                    )
+                    self.logger.info(test_log)
+
+        return losses.avg, top1.avg, top5.avg
+
+    def _train_epoch(self, epoch, optimizer, arch_train=False):
+        """
+        Train one epoch.
+        """
+        batch_time = AverageMeter("batch_time")
+        data_time = AverageMeter("data_time")
+        losses = AverageMeter("losses")
+        top1 = AverageMeter("top1")
+        top5 = AverageMeter("top5")
+
+        # switch to train mode
+        self.model.train()
+
+        data_loader = self.valid_loader if arch_train else self.train_loader
+        end = time.time()
+        for i, (images, labels) in enumerate(data_loader):
+            data_time.update(time.time() - end)
+            images = images.to(self.device, non_blocking=True)
+            labels = labels.to(self.device, non_blocking=True)
+
+            if self.arch_search:
+                perf_cost = Variable(
+                    torch.zeros(self.dev_num, 1), requires_grad=True
+                ).to(self.device, non_blocking=True)
+                output, perf_cost = self.model(images, self.temp, perf_cost)
+            else:
+                output = self.model(images)
+
+            loss = self.criterion(output, labels)
+
+            if self.arch_search:
+                # hardware-aware loss
+                regu_loss = self.reg_loss(perf_cost.mean(dim=0))
```

Comment:
I think regu_loss can be directly computed by trainer. See SPOS for an example.

Reply:
Meaning that the regu_loss is calculated for each candidate before training?

Reply:
The reg_loss is calculated with the perf_cost here, which is online sumed with the gumbel softmax coefficients and flops of each block at the same layer. Thanks:)

Reply:
I think `perf_cost` should also be calculated here.

Since you have all the mutators, you can learn from the mutators all the choices. And models only need to provide necessarily metadata to help mutators to sum flops.

See SPOS for an example.

Reply:
Done, the accumulation of perf_cost is moved outside of the forward function, and is conducted by the trainer._get_perf_cost().